### PR TITLE
Hardhat migration

### PR DESCRIPTION
**Pull request by wschwab**
This PR migrates the contracts from an Etherlime environment to a Hardhat one, including tests and deployment scripts.

All tests are currently passing with the exception of some tests on the EthBroker contract. There is also some other strange behaviour in those tests, which may have been introduced by use of the `hardhat_setBalance` call used to ensure testing accounts have sufficient ETH for the tests. Since not all tests are passing, the PR is marked as a draft.

The deployment script has been tested both locally and on a testnet.

### CODE REVIEW

**Comment by wschwab** [comment]
I've marked the PR as ready for review - the previous issues have been remedied by @RyRy79261 , and all tests are now passing.

Note that as of these changes the active folder for the contracts is no longer `bzzar-contracts/chain`, but rather `bzzar-contracts/packages/chain`. In order to run the test suite, navigate to this directory, run Yarn to install dependencies, and run `yarn hardhat test`.

File: LICENSE [hunk 1]
```diff
@@ -1,27 +0,0 @@
-Copyright (c) 2021 The Bzzaar Authors. All rights reserved.
```
Comment (significance) [inline LICENSE]: please retain the license

File: README.md [hunk 2]
```diff
@@ -1,245 +1,37 @@
...
-<img src="./docs/Swarm_Logo_Small.png">
- <h1>BZZ Smart Contract Ecosystem</h1>
-</div>
+# Bzzzar Hardhat Migration
 
----
+This repo has been set up to move the Bzzzar contracts from an Etherlime development environment to Hardhat, as Etherlime is no longer under active development. This was necessitated due to changes to the Eth Broker contract which caused the contract to longer pass its test suite in the Etherlime environment.
 
-# Index
+Due to the time-intensive nature of converting the test suite, at this juncture only the tests for the Eth Broker (`/test/broker.test.ts`) have been converted into the Hardhat-Waffle test format. Similarly, the deploy script is not yet converted to Hardhat.
```
Comment (significance) [inline README.md]: is this out of date? seems all the tests have been updated? 

not sure the commentary is required so can just be deleted ;D

File: README.md [hunk 3]
```diff
@@ -1,245 +1,37 @@
...
-    TOKEN_NAME=
-    TOKEN_SYMBOL=
-    # BZZ token has 16 decimals. DO NOT CHANGE THIS! This will break the bonding curve
-    TOKEN_DECIMAL=16
-    # BZZ needs this limit for the curve. DO NOT CHANGE THIS! This will break the bonding curve
-    TOKEN_CAP=1250000000000000000000000
-
-    # Testing address
-    ADDRESS_OF_TESTER=
+yarn test:chain ./test/broker.test.js
```
Comment (significance) [inline README.md]: i think this has been superseded too?

File: README.md [hunk 4]
```diff
@@ -1,245 +1,37 @@
...
 ```
+The other tests are all passing in the original Etherlime environment, even though we would like to migrate them to Hardhat too.
```
Comment (significance) [inline README.md]: similarly?

File: README.md [hunk 5]
```diff
@@ -1,245 +1,37 @@
...
+### Known Issues
 
-The testing portion of the deployment script will run when deploying to local or Rinkeby.
-
-The test deployment script will deploy the BZZ token, Bonding curve as well as a Mock DAI. The script will also mint the test user address some Mock DAI. This speeds up the testing process.
-
-Please ensure that you have filled in all the following values from the `.env.example` before attempting to run the script:
+#### Running out of funds
 
+The deployer account runs out of Eth if the Eth Broker tests are run all at once. Our current solution until we can find the cause in the tests is to run the tests one at a time, and to restart the chain after each test. You can single out a particular test by tagging it with `only`, for example:
```
Comment (significance) [inline README.md]: this is fixed too?

File: README.md [hunk 6]
```diff
@@ -1,245 +1,37 @@
...
+```javascript
+it.only("burn balance checks", async () => {
 ```
-# Deployer EOA private key (insecure)
```
Comment (significance) [inline README.md:134]: am wondering has this been removed?

File: packages/chain/.example.env [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+# To use, copy this file, rename it .env, and fill in the values
+# Everything will be interpreted as a string by JavaScript even though 
+# you should not wrap values in quotation marks
+
+# Infura endpoints should be passed in WITHOUT the rest of the url
+# Private keys should be pasted WITHOUT the 0x prefix
+
+# Example
+EXAMPLE_INFURA_KEY=582dabbadbeef8...
+EXAMPLE_DEPLOYER_PRIV_KEY=deadbeef01EEdf972aBB...
```
Comment (significance) [inline packages/chain/.example.env:13]: no d34db33s! 😱

File: packages/chain/artifacts/@openzeppelin/contracts/GSN/Context.sol/Context.dbg.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (significance) [inline packages/chain/artifacts/@openzeppelin/contracts/GSN/Context.sol/Context.dbg.json]: this stuff is included in gitignore so should not be committed to repo?

File: packages/chain/contracts/Eth_broker.sol [hunk 9]
```diff
@@ -217,7 +216,7 @@ contract Eth_broker {
         mutex()
         returns (bool)
     {
-        (uint256 daiNeeded, uint256 ethReceived) = _commonMint(
+        (uint256 daiNeeded, uint256 ethReceived, uint256 ethSpent) = _commonMint(
```
Comment (significance) [inline packages/chain/contracts/Eth_broker.sol]: does this PR also fix bugs? my understanding is that it only upgrades the development environment. it would be better to keep bug fixes in separate PR's unless there is a compelling reason to do otherwise?

File: packages/chain/scripts/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
+# Warp Core Deploy Script
```
Comment (significance) [inline packages/chain/scripts/README.md]: please remove this or rewrite it to be repo specific

File: packages/chain/scripts/orig_deploy.js [hunk 11]
```diff
@@ -20,7 +20,7 @@ const defaultConfigs = {
   gasLimit: ethers.utils.parseUnits(process.env.GAS_LIMIT, 0),
   gasPrice: ethers.utils.parseUnits(process.env.GAS_PRICE, 0),
   chainId: 1,
-	etherscanApiKey: ""
+	etherscanApiKey: "3DQYBPZZS77YDR15NKJHURVTV9WI2KH6UY"
```
Comment (significance) [inline packages/chain/scripts/orig_deploy.js]: should be in env vars

File: packages/chain/artifacts/@openzeppelin/contracts/GSN/Context.sol/Context.dbg.json (hunk 8 above)
Comment (ralph-pichler) [reply packages/chain/artifacts/@openzeppelin/contracts/GSN/Context.sol/Context.dbg.json]: yes, please remove `packages/chain/artifacts` and `packages/chain/cache` from the repo.

File: packages/chain/scripts/README.md (hunk 10 above)
Comment (ralph-pichler) [reply packages/chain/scripts/README.md]: anyway this is not important as we use different tooling for deployment.

**Comment by ralph-pichler** [comment]
getting a lot of errors like
```
test/broker.test.ts:63:16 - error TS2345: Argument of type 'SignerWithAddress' is not assignable to parameter of type 'string | Provider | Signer'.
  Property 'getFeeData' is missing in type 'SignerWithAddress' but required in type 'Signer'.

63       .connect(owner)
```
when trying to run tests.

**Comment by ralph-pichler** [comment]
> getting a lot of errors like
> 
> ```
> test/broker.test.ts:63:16 - error TS2345: Argument of type 'SignerWithAddress' is not assignable to parameter of type 'string | Provider | Signer'.
>   Property 'getFeeData' is missing in type 'SignerWithAddress' but required in type 'Signer'.
> 
> 63       .connect(owner)
> ```
> 
> when trying to run tests.

oddly enough this does not happen when running tests in ci environment. so maybe a local issue on my end.

**Comment by significance** [comment]
> > getting a lot of errors like
> > ```
> > test/broker.test.ts:63:16 - error TS2345: Argument of type 'SignerWithAddress' is not assignable to parameter of type 'string | Provider | Signer'.
> >   Property 'getFeeData' is missing in type 'SignerWithAddress' but required in type 'Signer'.
> > 
> > 63       .connect(owner)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > when trying to run tests.
> 
> oddly enough this does not happen when running tests in ci environment. so maybe a local issue on my end.

hmm, strange, tests pass locally for me too

File: packages/chain/scripts/README.md (hunk 10 above)
Comment (significance) [reply packages/chain/scripts/README.md]: it would be good to provide either this functionality or expose the other deployment tools for other users of the library

File: README.md (hunk 2 above)
Comment (wschwab) [reply README.md]: The tests and deploy script have been updated, I'll remove the outdated parts.

File: packages/chain/.example.env (hunk 7 above)
Comment (wschwab) [reply packages/chain/.example.env:13]: Why would I kill the bees when I can kill the beef? :smile:

File: packages/chain/artifacts/@openzeppelin/contracts/GSN/Context.sol/Context.dbg.json (hunk 8 above)
Comment (wschwab) [reply packages/chain/artifacts/@openzeppelin/contracts/GSN/Context.sol/Context.dbg.json]: will do

File: packages/chain/contracts/Eth_broker.sol (hunk 9 above)
Comment (wschwab) [reply packages/chain/contracts/Eth_broker.sol]: The reason for the migration was due to shortcomings in the original environment (since Etherlime is no longer maintained). While I do agree that it is theoretically better to split into two PRs, this should have been done some time ago when the original migration was made. At this point my personal opinion would be that it's making extra labour for little gain.

If you are very definite about wanting to do this, I should be able to roughly simulate it by breaking out all of the work from the last commit and re-adding it back in, but I find that to be of very limited value.

File: packages/chain/scripts/orig_deploy.js (hunk 11 above)
Comment (wschwab) [reply packages/chain/scripts/orig_deploy.js]: This is the old deploy script, I'll just remove it, but you are correct that RIP on that key.

File: packages/chain/contracts/Eth_broker.sol (hunk 9 above)
Comment (significance) [reply packages/chain/contracts/Eth_broker.sol]: really sorry dude but i'm going to have to insist on this because i will have to write a risk assessment for the changes and it will need to be attached to a PR with only those changes involved. sorry, i know it's annoying 😌

File: packages/chain/scripts/README.md (hunk 10 above)
Comment (wschwab) [reply packages/chain/scripts/README.md]: I'd like a confirmation on whether I should leave in as-is, or do something else.

**Comment by wschwab** [comment]
As requested, I 've split out the working Eth Broker tests in order to split out the Hardhat migration and fixing the tests into separate PRs. Most code review comments here are more relevant to the Eth Broker PR imho, though I have tried to update the README in this PR to reflect the state of the repo. If you would like changes to the deploy script README specifically in this PR, please let me know. Otherwise I believe this PR is ready for merge.

**Comment by significance** [comment]
> As requested, I 've split out the working Eth Broker tests in order to split out the Hardhat migration and fixing the tests into separate PRs. Most code review comments here are more relevant to the Eth Broker PR imho, though I have tried to update the README in this PR to reflect the state of the repo. If you would like changes to the deploy script README specifically in this PR, please let me know. Otherwise I believe this PR is ready for merge.

thanks for this dude - much appreciated!

**Review by ralph-pichler — CHANGES_REQUESTED** [review]
artifacts file and cache are still in the directory. also it seems the changes from eth broker contract are still in this PR. also please add the commit from https://github.com/wschwab/bzzaar-contracts/pull/2 to this branch.

File: README.md (hunk 3 above)
Comment (wschwab) [reply README.md]: this command should still be true, I figured it might be useful for running specifically the old Broker tests in isolation

**Review by ralph-pichler — CHANGES_REQUESTED** [review]
the ci still is not updated. please add the workflow file from https://github.com/wschwab/bzzaar-contracts/pull/2

**Comment by wschwab** [comment]
> the ci still is not updated. please add the workflow file from [wschwab#2](https://github.com/wschwab/bzzaar-contracts/pull/2)

This should be fixed now, I missed it before.

**Review by ralph-pichler — APPROVED** [review]
looks good now. tests on broker are failing but that was to be expected.

File: packages/chain/test/curve_calc.test.js [hunk 12]
```diff
@@ -180,26 +196,30 @@ describe("🧮 Curve Calculations Tests", () => {
...
-                test_settings.dai.curve_coll_at_prem,
-                "Prim function for post-buy incorrect"
-            );
-            assert.equal(
-                spotPricePostMint.toString(),
-                test_settings.dai.one_cost,
-                "Start cost per token incorrect"
-            );
+            expect(buyCost.toString()).to.equal(test_settings.dai.buyCost);
+            // assert.equal(
```
Comment (significance) [inline packages/chain/test/curve_calc.test.js]: can you remove the comments here please dude

File: packages/chain/test/curve_calc.test.js [hunk 13]
```diff
@@ -180,26 +196,30 @@ describe("🧮 Curve Calculations Tests", () => {
...
+            //     test_settings.dai.buyCost,
+            //     "Mint cost incorrect"
+            // );
+            expect(primFuncAtZero.toString()).to.equal(pre_mint_sequence.dai.cost);
+            // assert.equal(
+            //     primFuncAtZero.toString(),
+            //     pre_mint_sequence.dai.cost,
+            //     "Prim function for pre-buy incorrect"
+            // );
+            expect(primFuncAtPreMint.toString()).to.equal(test_settings.dai.curve_coll_at_prem);
```
Comment (significance) [inline packages/chain/test/curve_calc.test.js:172]: can you fix indentation please?

File: packages/chain/scripts/README.md (hunk 10 above)
Comment (wschwab) [reply packages/chain/scripts/README.md]: I've removed in the meantime

**Comment by wschwab** [comment]
all conversations should be resolved now, feel free to give another look and see if there's anything I missed

**Review by significance — APPROVED** [review]
